Change: no longer use UDP when entering the lobby of a server

TrueBrain opened the pull request:
## Motivation / Problem

With STUN in mind, it is a lot easier if we have a single protocol to talk between client and server. This was mostly true, except for a part in the lobby window.

Most likely to avoid code-duplication (random guess honestly), opening the lobby does two requests: one via UDP to get the latest game-info, and one via TCP to get the latest company-info. This PR sets out to change that both to TCP.

Accidentally, this solves a rare edge-case, where someone create a new company on a server, TCP reacts with the company-list, but the UDP packet is lost. In this case, the new company is not shown. Only until the UDP packet is received, it will be shown correctly. By using TCP here over UDP, we ensure this isn't possible.

## Description

The first commit moves a lot of code related to NetworkGameInfo to a single place, so adding GameInfo to TCP (next to UDP) doesn't cause code duplication.

The second commit adds a packet to the TCP to exchange GameInfo over TCP too.

---

Sadly, I wasn't a very smart cookie when I wrote the network protocol, and over TCP we do not handshake a protocol-version before we initiate conversation. This means that there are potential (future) issues with this PR. Currently this isn't an issue, but just to walk through the situations:

New client + new server:
- MasterServer tells the servers online, via UDP these are queried. When pressing Join, the Lobby window does everything via TCP.
- Via "Add Server" someone manually adds a server. Via UDP it is queried. When pressing Join, the Lobby window does everything via TCP.

Old client + new server / new client + old server:
- MasterServer tells the servers online, via UDP these are queried. Versions are incompatible, so lobby cannot be opened.
- Via "Add Server" someone manually adds a server. Via UDP it is queried. Versions are incompatible, so lobby cannot be opened.

Using `-n` to join a server bypasses lobby, and as such works fine too (well, it hits revision mismatch if that is the case, ofc)

## Backport

Currently "Add Server" still uses UDP to query the server. A future PR will change this to also use TCP here (only for manual servers and manual refresh). This does mean that adding older servers will result in a protocol error, as older server don't know this new packet. In such future PR, the client will pick up on this and show a nice "Server is too old" text.

It would help to backport this to 1.11. It would mean that when 1.12 releases, the 1.11.N can be added manually and show the game info (and that the version is incompatible). This would mean for most of our servers there isn't a real problem, and it is only an issue if people add really old servers manually.
This is also the reason that this future PR is not included in this PR.

Additionally, the last two commits from this PR don't have to be backported. It makes use of this new TCP packet; without it nothing breaks, and means there is no functional difference after backporting. But it is better to play it safe, so do not backport the last two commits please.

Additionally, if this is backported, so should #9129, as it fixes packet order.

## Limitations

None, this should be fully backports compatible. The GUI prevents opening the lobby for old servers, and UDP is used for anything before that.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

TrueBrain committed:
Codechange: move all NetworkGameInfo related functions to a single file

It currently was a bit scattered over the place. Part of
NetworkGameInfo is also the GRF Identifiers that goes with it.

rubidium42 commented on src/network/core/game_info.cpp:
```diff
@@ -106,6 +106,22 @@ bool IsNetworkCompatibleVersion(const char *other)
+	}
+
+	/* Check if we are allowed on this server based on the revision-match */
+	ngi.version_compatible = IsNetworkCompatibleVersion(ngi.server_revision);
```
I would move this code to the begin of the function, and set ngi.compatible to ngi.version_compatible. Then the if below is not needed anymore.

rubidium42 commented on src/network/network_client.cpp:
```diff
@@ -341,15 +342,21 @@ static_assert(NETWORK_SERVER_ID_LENGTH == 16 * 2 + 1);
-	Packet *p = new Packet(PACKET_CLIENT_COMPANY_INFO);
-	my_client->SendPacket(p);
+	Packet *p1 = new Packet(PACKET_CLIENT_GAME_INFO);
+	my_client->SendPacket(p1);
```
`my_client->SendPacket(new Packet(PACKET_CLIENT_GAME_INFO));` achieves the same without the (badly named) local variable.

rubidium42 commented on src/network/core/game_info.cpp:
```diff
@@ -0,0 +1,323 @@
+	 * The parts must be read in the same order as they are sent!
+	 */
+
+	/* Update the documentation in udp.h on changes
```
Reference is broken by your move of this documentation.

rubidium42 commented on src/network/core/game_info.cpp:
```diff
@@ -0,0 +1,323 @@
+	 * The parts must be read in the same order as they are sent!
+	 */
+
+	/* Update the documentation in udp.h on changes
```
Reference is broken by your move of this documentation.

rubidium42 commented on src/network/network_udp.cpp:
```diff
@@ -171,29 +172,10 @@ void ServerNetworkUDPSocketHandler::Receive_CLIENT_FIND_SERVER(Packet *p, Networ
-
-	strecpy(ngi.server_name, _settings_client.network.server_name, lastof(ngi.server_name));
-	strecpy(ngi.server_revision, GetNetworkRevisionString(), lastof(ngi.server_revision));
+	FillNetworkGameInfo(ngi);
```
All calls to this function are followed by the SendNetworkGameInfo function, without modification to the data, right?
So why have a local variable of the data to pass to SendNetworkGameInfo directly after it?.

TrueBrain replied:
Mostly as I think `FillNetworkGameInfo` is a bullshit function and should be removed, but that I didn't want to include in this PR. I cannot believe we are, for every client request, fetching this almost static information. We should just have a static version ready to go, honestly.

So I left it in two functions so that future work is easier.

rubidium42 commented on src/network/network_udp.cpp:
```diff
@@ -171,29 +172,10 @@ void ServerNetworkUDPSocketHandler::Receive_CLIENT_FIND_SERVER(Packet *p, Networ
 
 	Packet packet(PACKET_UDP_SERVER_RESPONSE);
-	this->SendNetworkGameInfo(&packet, &ngi);
+	SendNetworkGameInfo(&packet, &ngi);
```
The name is somewhat misleading, yes it was in the past too. It's not actually sending the GameInfo, it's adding it to the packet. Though it does somewhat conform to the functions in Packet, so I'm in a bit of doubt whether it is an issue or not.

TrueBrain replied:
I agree. The naming comes from these pairs:

```
void ReceiveGRFIdentifier(Packet *p, GRFIdentifier *grf);
void SendGRFIdentifier(Packet *p, const GRFIdentifier *grf);

void ReceiveNetworkGameInfo(Packet *p, NetworkGameInfo *info);
void SendNetworkGameInfo(Packet *p, const NetworkGameInfo *info);
```

Neither are really sending. I could prefix them with `Prepare` or postfix with `Packet` or something, but either they should all change or none, in my opinion. But neither the prefix or postfix sounds good to me .. so I am still looking for a good name :D

TrueBrain replied:
`Receive` -> `Deserialize` and `Send` -> `Serialize`? 

Edit: yeah, I went for this. The doxygen comment even says that is what it is doing :D

rubidium42 commented on src/network/core/game_info.h:
```diff
@@ -50,6 +89,19 @@ struct NetworkGameInfo : NetworkServerGameInfo {
+void ReceiveNetworkGameInfo(Packet *p, NetworkGameInfo *info);
+void SendNetworkGameInfo(Packet *p, const NetworkGameInfo *info);
+
+void HandleIncomingNetworkGameInfoGRFConfig(GRFConfig *config);
```
Is only used in game_info.cpp, so why make it a "public" function, just make it a static function in game_info.cpp.

rubidium42 commented on src/network/core/game_info.h:
```diff
@@ -50,6 +89,19 @@ struct NetworkGameInfo : NetworkServerGameInfo {
-const char * GetNetworkRevisionString();
+extern NetworkServerGameInfo _network_game_info;
+
+const char *GetNetworkRevisionString();
```
I'm not sure whether version handling belongs with the game info. At some level it makes sense, on the other hand it's used more generic in other places of the network code.

TrueBrain replied:
I had the same consideration, but in the end, I decided it is at least not unreasonable to be here, and as it was already here, not to move it.

TrueBrain committed:
Add: ability to retrieve game info from server over TCP

TrueBrain committed:
Codechange: refactor CheckGameCompatibility() from existing function

Later commits use this function in other places too.

TrueBrain committed:
Change: no longer use UDP when entering the lobby of a server

The lobby of a server requested some parts via UDP and some via
TCP. This is strictly seen fine, but for future extensions it
is a lot easier if just one protocol is used.

PeterN commented on src/network/core/game_info.cpp:
```diff
@@ -106,6 +106,21 @@ bool IsNetworkCompatibleVersion(const char *other)
+/**
+ * Check if an game entry is compatible with our client.
+ */
+void CheckGameCompatability(NetworkGameInfo &ngi)
```
Should be spelled Compatibility

PeterN commented on src/network/core/game_info.cpp:
```diff
@@ -0,0 +1,323 @@
+		}
+
+		/* Prepare a prefix of the git hash.
+		* Size is length + 1 for terminator, +2 for -g prefix. */
```
I know this is moved code, but missing a space before the comment here.

rubidium42 commented on src/network/core/game_info.cpp:
```diff
@@ -0,0 +1,323 @@
+
+	const char *hash1 = ExtractNetworkRevisionHash(GetNetworkRevisionString());
+	const char *hash2 = ExtractNetworkRevisionHash(other);
+	return hash1 && hash2 && (strncmp(hash1, hash2, GITHASH_SUFFIX_LEN) == 0);
```
If you have to touch this commit due to a conflict, can you add `!= nullptr` after the first `hash1` and `hash2`?

TrueBrain committed:
Codechange: be explicit in pointer comparisons
